Add 'Discounts Using Lists' chapter (#38)
melange-re/melange-for-react-devs: docs/discounts-lists/index.md

# Discounts Using Lists

You've implemented the logic for the burger discounts that will be applied on
International Burger Day, but something feels off. The code inside the
`Discount` module is safe, in the sense that it doesn't raise exceptions, so
when it's used in the app, it won't crash. But the current logic in
`Discount.getFreeBurger` is a little fragile, since you have to remember to not
change the order of function invocations in such a way that the function will
[accidentally change the input array](/burger-discounts/#arrays-are-mutable).
You have a test in `DiscountTests` which tests for that possibility, but you
know that tests are only as reliable as the humans that maintain them. By now,
you've come to realize that The OCaml Way should be rewrite the function so that

> it **cannot** have side effects.

Reviewer: It's cool to help readers reason about this refactor to use lists instead of arrays based on mutation / side effects and the desire to avoid them.

But I wonder if it'd be necessary to mention as well that the `getFreeBurger` code didn't require any access to random positions in a collection of values, which is what arrays excel at. And that the absence of this requirement is what enables us to transition to lists more easily?

Author: Added a tip about that in https://github.com/melange-re/melange-for-react-devs/pull/38/commits/75288343071a655000c32ac8de9277e1137dc2cf

through the [Stdlib.List
module](https://melange.re/v3.0.0/api/re/melange/Stdlib/List/), which has many
functions that are equivalent to the ones you've already used in
[Js.Array](https://melange.re/v3.0.0/api/re/melange/Js/Array/).

::: tip

Since `Stdlib` is opened by default, you can access all `Stdlib.List` functions
using just `List`, e.g. `List.map`, `List.filter`, etc. Because of this, we'll
refer to this module as `List` from now on.

:::

## Refactor `Discount.getFreeBurger`

Let's refactor `Discount.getFreeBurger` to accept `list(Item.t)` instead of
`array(Item.t)`:

<<< Discount.re#get-free-burger

There are a lot of things to talk about in this piece of code---let's go through
them one by one.

## Documentation comment

You might have noticed that the comment at the beginning of the function has a
different format than before:

```reason
// Buy 2 burgers, get 1 free // [!code --]
/** Buy 2 burgers, get 1 free */ // [!code ++]
```

This is a **documentation comment**, a special comment that is attached to the
function it appears above. Go to `DiscountTests` and hover over an invocation of
`getFreeBurger`---the editor will display a popup showing both the type

> signature and documentation comment of this function.

Reviewer: Maybe show a small snapshot of how it looks? e.g.

<img width="331" alt="image" src="https://github.com/melange-re/melange-for-react-devs/assets/220424/5191ff9f-18db-465e-95b1-93b22da7a733">

Author: Added screenshot in https://github.com/melange-re/melange-for-react-devs/pull/38/commits/68a6517dffba6de1b470128dc782b080751ca6ae

  Note that `List.filter` doesn’t accept the labeled argument `~f`, because the
  functions inside `List` don't use labeled arguments.
- `Js.Array.map` →
  [List.map](https://melange.re/v3.0.0/api/re/melange/Stdlib/List/#val-map).
  `List.map` also doesn’t accept the labeled argument `~f`.
- `Js.Array.sortInPlaceWith` →
  [List.sort](https://melange.re/v3.0.0/api/re/melange/Stdlib/List/#val-sort).
  `List.sort` returns a brand new list, because, unlike
`Js.Array.sortInPlaceWith`, it doesn’t modify its argument (and it can’t, since
  lists are immutable).

## Pattern match on lists

The switch expression in `Discount.getFreeBurgers` accepts the entire `prices`
list. Unlike with arrays, we can pattern match on lists even if we don’t know
the length of the list.

Inside the "failure" branch of the switch expression, we see this:

```reason
| []
| [_] => None
```

This pattern matches both empty lists and lists with one element, and returns
`None`.

Inside the "success" branch of the switch expression, we have:

```reason
| [_, cheaperPrice, ..._] => Some(cheaperPrice)
```

This pattern will match on lists that have at least two elements. The first

> element is ignored via `_`. The second element is bound to the name

Reviewer: ```suggestion
element is ignored via the wildcard pattern `_`. The second element is bound to the name
```

Author: Improved in https://github.com/melange-re/melange-for-react-devs/pull/38/commits/cfe9cc81174c88c8345ccc03cd8cba25739abfc5

`cheaperPrice`, which is encased in `Some` and returned. We use *list spread

> syntax* (`...`) to indicate that the list can have more than the two elements we

Reviewer: Maybe worth mentioning the spread operator for lists is a feature exclusively available in Reason syntax? https://reasonml.github.io/docs/en/basic-structures#list

Author: From my perspective, it's just the Reason equivalent of the `::` operator in native syntax. Or is there something special about it that I'm missing?

Reviewer: No, that's ok.

explicitly matched on.

## List spread syntax

> The spread operator (`...`) is also used to add elements to the front of a list:

Reviewer: This wording had me confused. Maybe because we're not adding elements to a list but rather creating a new one?

```suggestion
The spread operator (`...`) can also be used to create a new list by prepending
elements to an existing list:
```

Author: Improved in https://github.com/melange-re/melange-for-react-devs/pull/38/commits/cfe9cc81174c88c8345ccc03cd8cba25739abfc5

```reason
let list = [1, 2, 3];
Js.log([0, ...list]); // [0, 1, 2, 3]
Js.log([-1, 0, ...list]); // [-1, 0, 1, 2, 3]
```

::: tip

```reason
[1, 2, 3]
```

is really just a shortcut for

```reason
[1, ...[2, ...[3, ...[]]]]
```

:::

When pattern matching, the spread operator allows you to bind the **tail** of
a list to a name:

```reason
switch ([1, 2, 3]) {
| [_, ...tail] => Js.log(tail) // [2, 3]
| _ => ()
};

switch ([1, 2, 3]) {
| [_, _, _, ...tail] => Js.log(tail) // []
| _ => ()
};
```

The tail of the list is the sublist that remains after you extract the first n
elements from the front of the list. As you can see, the tail might be the empty
list (`[]`). In practice, you don't need to bind the tail to a name unless

> you're writing a [custom list function](/todo). Often, you'll just bind the tail

Reviewer: Is this todo part of this PR?

Author: No, I have tentative plans to create another chapter that talks about creating your own list functions using recursion. But I haven't decided yet, and anyway it hasn't been written yet.

> to `_`, effectively ignoring it:

Reviewer: ```suggestion
to the wildcard pattern `_`, effectively ignoring it:
```

Author: Improved in https://github.com/melange-re/melange-for-react-devs/pull/38/commits/cfe9cc81174c88c8345ccc03cd8cba25739abfc5

```reason{2}
switch (["one", "two", "three"]) {
| [a, b, ..._] => Js.log(a ++ ", " ++ b ++ ", etc") // one, two, etc
| _ => ()
};
```

## Runtime representation of lists

If you run the code snippets above in the [Melange
Playground](https://melange.re/v3.0.0/playground/?language=Reason&code=bGV0IGxpc3QgPSBbMSwgMiwgM107CkpzLmxvZyhbMCwgLi4ubGlzdF0pOyAvLyBbMCwgMSwgMiwgM10KSnMubG9nKFstMSwgMCwgLi4ubGlzdF0pOyAvLyBbLTEsIDAsIDEsIDIsIDNdCgpzd2l0Y2ggKFsxLCAyLCAzXSkgewp8IFsxLCAuLi50YWlsXSA9PiBKcy5sb2codGFpbCkgLy8gWzIsIDNdCnwgXyA9PiAoKQp9OwoKc3dpdGNoIChbMSwgMiwgM10pIHsKfCBbMSwgMiwgMywgLi4udGFpbF0gPT4gSnMubG9nKHRhaWwpIC8vIFtdCnwgXyA9PiAoKQp9Owo%3D&live=off),

> you'll see output like this:
> 
> ```text
> {"hd":0,"tl":{"hd":1,"tl":{"hd":2,"tl":{"hd":3,"tl":0}}}}
> {"hd":-1,"tl":{"hd":0,"tl":{"hd":1,"tl":{"hd":2,"tl":{"hd":3,"tl":0}}}}}
> {"hd":2,"tl":{"hd":3,"tl":0}}

Reviewer: I wonder if it'd help to show this output formatted, to clarify how it works, e.g.:

```suggestion
you'll see output like this (formatted):

```text
{
  "hd": 0,
  "tl": {
    "hd": 1,
    "tl": {
      "hd": 2,
      "tl": {
        "hd": 3,
        "tl": 0
      }
    }
  }
}
{
  "hd": -1,
  "tl": {
    "hd": 0,
    "tl": {
      "hd": 1,
      "tl": {
        "hd": 2,
        "tl": {
          "hd": 3,
          "tl": 0
        }
      }
    }
  }
}
{
  "hd": 2,
  "tl": {
    "hd": 3,
    "tl": 0
  }
}
```

Author: Improved in https://github.com/melange-re/melange-for-react-devs/pull/38/commits/3de17f2fc834b3a4305568f218f271af01bc06e3

       switch (item) {
       | Item.Burger(_) => true
       | Sandwich(_)
       | Hotdog => false
       }
     )
  |> List.map(Item.toPrice)
  |> List.sort((x, y) => - compare(x, y))
  |> List.nth_opt(1);
};
```

By using `List.nth_opt`, we can simplify the function to a single expression.
However, we get a compilation error:

```text
File "docs/order-confirmation/Discount.re", line 61, characters 18-19:
61 |   |> List.nth_opt(1);
                       ^
Error: This expression has type int but an expression was expected of type
         'a list
```

## Placeholder operator

This is because of the type signature of `List.nth_opt` is

```text
list('a) => int => option('a)
```

That is, it accepts the list as the first argument, not the last. Recall that
the pipe last operator (`|>`) pipes values into the last argument of a function.
However, there's a way to override the placement of the argument:

```reason
|> List.nth_opt(_, 1)
```

Do not confuse the `_` here for wildcard, here it's a placeholder for where the
argument should go. When we put `_` in the first argument position, it overrides
the default behavior of the pipe last operator. Run `npm run test` to confirm

> that `Discount.getFreeBurger` works the same as before.

Reviewer: I personally find this wildcard operator very hacky because it's only available in Reason syntax (so it's not possible to convert to OCaml). Maybe we could just get the result from the code above in a `prices` list? Like it had before. Then just leave `List.nth_opt(prices, 1)` at the end.

That way, we skip the compilation error, and the need for this whole section to explain some functionality that is not really "The OCaml Way" :)

Author: Yeah, it's not very OCaml Way, but I decided to be a little pragmatic and write about this because I anticipate talking about Belt in a later chapter. At the moment, using the placeholder operator is the only way to use most of the Belt API with the pipe last operator. I don't feel very strongly about it though, I could certainly just move this section into whatever future chapter actually brings up Belt. Wdyt?

maintainable, and you've also learned a lot about lists along the way. In the
next chapter, we'll finally use the discount logic to reduce the final price of
an order.

## Overview

- Lists are immutable
- You can pattern match on a whole list, even if you don't know its length
- The `List` module contains most of the functions you'll need for
  dealing with lists
- The runtime representation of lists:
  - Empty list → `0`
  - Nonempty list → a JavaScript object with the fields `hd` (for head) and `tl`
    (for tail)
- The delimiters for list literals are `[]`
- You can use list spread syntax (`...`) to add elements to the front of a list
  or pattern match on the tail of a list
  - The names of equivalent functions might not match the names in `Js.Array`
- Documentation comments show up in editor hover popups and generated
  documentation pages
  - They can be attached to functions, modules, types, and
  variables
- The `ListLabels` module contains the same functions as in `List`, but they
  have labeled arguments instead of positional arguments
- The placeholder operator (`_`) can be used to override the position of the

>   piped argument when using the pipe last operator

Reviewer: If you decide to follow the comment above, this section would have to be removed.

## Exercises

<b>1.</b> There are a couple ways to improve `Discount.getFreeBurger`:

- Instead of `StdLib.compare`, use a type-specific compare function. Examples
  of a type-specific compare functions are `Bool.compare` or `String.compare`.
- Use

>   [List.filter_map](https://melange.re/v2.2.0/api/re/melange/Stdlib/List/#val-filter_map)

Reviewer: This exercise is not explaining the problem that has to be solved (improve the accuracy of comparisons, I assume) but rather going to the solution directly (use other `compare` functions).

Author: Yeah, I feel I've always been told to use the type-specific compare function but I can't find a quote anywhere to support that advice. Maybe I should just delete the part about changing it to `Float.compare`?

Reviewer: Maybe you could explain the issues with `Stdlib.compare` first? E.g.

> While `Stdlib.compare` can be handy as it can be used with multiple types, its flexibility comes with some drawbacks. It can be slower due to its polymorphic nature, and might not always offer meaningful comparisons for complex types. Additionally, it can raise exceptions if used on non-comparable types (like functions). As a result, for domain-specific or performance-critical applications, defining custom comparison functions is often a safer and more efficient approach.

Author: Added an explanation in https://github.com/melange-re/melange-for-react-devs/pull/38/commits/f9f936f5b2c00c50dc6d50ac1c9dc46a0d873691

::: details Solution

<<< Discount.re#get-free-burger-improved

Prefer type-specific compare functions like [Float.compare](https://melange.re/v1.0.0/api/re/melange/Stdlib/Float/#val-compare) over polymorphic
`compare`.

:::

<b>2.</b> Add a safe `Lst.nth` function which returns the nth element of a list
encased in `Some`. If the nth element doesn’t exist, return `None`. Make sure
the function can be used with the pipe last operator without the use of the
placeholder operator. Refactor `Discount.getFreeBurger` to use your new
function.

::: tip

> Note that `Lst` is not a typo. If we name our new module `List`, it will shadow

Reviewer: Maybe call it `EnhancedList`? Or `SafeList`? That way the tip can be skipped.

Author: Renamed in https://github.com/melange-re/melange-for-react-devs/pull/38/commits/e9a4db32b0d9354f8a4455b60a7723f3b113767a
